api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java

```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+	 * @param includeVoided true if voided person should be included 
+	 * @param gender of the person to search for 
+	 * @return the set of Persons that match the search criteria
+	 */
+	private Set<Person> executeSoundexTwoPersonNameQuery(String name1, String name2,  Integer birthyear, boolean includeVoided , String gender) {
+		// Use the Soundex Search Mode for each query
+		String matchMode = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE);
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE, OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_SOUNDEX);
+		
+		PersonLuceneQuery personLuceneQuery = new PersonLuceneQuery(sessionFactory);
+		int maxResults = HibernatePersonDAO.getMaximumSearchResults();
+		
+		// Execute a search on all names of a person with the given criteria
+		LuceneQuery<PersonName> luceneQuery = personLuceneQuery.getSoundexPersonNameSearchOnAllNames(name1, name2, 0, false, "M");
+		LinkedHashSet<Person> people = new LinkedHashSet<>();
+		ListPart<Object[]> names = luceneQuery.listPartProjection(0, maxResults, "person.personId");
+		names.getList().forEach(x -> people.add(getPerson((Integer) x[0])));
+		
+		Set<Person> firstNamesMatch = new HashSet<>();
+		Set<Person> secondNamesMatch = new HashSet<>();
+		Set<Person> secondNamesMatchFirstName = new HashSet<>();
+		Set<Person> firstNamesMatchSecondName = new HashSet<>();
+		
+		
+		// Executed Filter Logic: 
+		// People whose first name matches n1 and second part of the name (givenName, familyName, familyName2) matches at least once n2
+		firstNamesMatch = execSoundexLuceneQueryByNames(true, name1, birthyear, includeVoided, gender);
+		secondNamesMatch = execSoundexLuceneQueryByNames(false, name2, birthyear, includeVoided, gender);
+		
+		firstNamesMatch.retainAll(secondNamesMatch);
+		people.addAll(firstNamesMatch);
+		
+		// Executed Filter Logic: 
+		// People who have second part of the name(givenName, familyName, familyName2) matching n2 and n1 (at least once)
+		secondNamesMatchFirstName = execSoundexLuceneQueryByNames(false, name1, birthyear, includeVoided, gender);
+		secondNamesMatch.retainAll(secondNamesMatchFirstName);
+		people.addAll(secondNamesMatch);
+		secondNamesMatch.clear();
+		
+		// Executed Filter Logic:
+		// People who have given_name matching n2, second name matching n1, and another part of second name matching “”
+		firstNamesMatchSecondName = execSoundexLuceneQueryByNames(true, name2, birthyear, includeVoided, gender);
+		secondNamesMatch = execSoundexLuceneQueryByNames(false, "", birthyear, includeVoided, gender);
+		
+		firstNamesMatchSecondName.retainAll(secondNamesMatchFirstName);
+		firstNamesMatchSecondName.retainAll(secondNamesMatch);
+		people.addAll(firstNamesMatchSecondName);
+		
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE, matchMode);
```

Comment:
What happens when two people search for patients at the same time in different browsers?